commands/publish.go

```diff
@@ -0,0 +1,228 @@
+
+	// Setup flags that are used only by this command (variables defined above)
+	publishCmd.Flags().BoolVar(&nocache, "no-cache", false, "Do not use Docker's build cache")
+	publishCmd.Flags().BoolVar(&squash, "squash", false, `Use Docker's squash flag for smaller images [experimental] `)
+	publishCmd.Flags().IntVar(&parallel, "parallel", 1, "Build in parallel to depth specified.")
+	publishCmd.Flags().BoolVar(&shrinkwrap, "shrinkwrap", false, "Just write files to ./build/ folder for shrink-wrapping")
+	publishCmd.Flags().StringArrayVarP(&buildArgs, "build-arg", "b", []string{}, "Add a build-arg for Docker (KEY=VALUE)")
+	publishCmd.Flags().StringArrayVarP(&buildOptions, "build-option", "o", []string{}, "Set a build option, e.g. dev")
+	publishCmd.Flags().Var(&tagFormat, "tag", "Override latest tag on function Docker image, accepts 'latest', 'sha', 'branch', or 'describe'")
+	publishCmd.Flags().StringArrayVar(&buildLabels, "build-label", []string{}, "Add a label for Docker image (LABEL=VALUE)")
+	publishCmd.Flags().StringArrayVar(&copyExtra, "copy-extra", []string{}, "Extra paths that will be copied into the function build context")
+	publishCmd.Flags().BoolVar(&envsubst, "envsubst", true, "Substitute environment variables in stack.yml file")
+	publishCmd.Flags().BoolVar(&quietBuild, "quiet", false, "Perform a quiet build, without showing output from Docker")
+	publishCmd.Flags().BoolVar(&disableStackPull, "disable-stack-pull", false, "Disables the template configuration in the stack.yml")
+	publishCmd.Flags().StringVar(&platforms, "platforms", "linux/amd64", "A set of platforms to publish")
+	publishCmd.Flags().StringArrayVar(&extraTags, "extra-tag", []string{""}, "Additional extra image tag")
+
+	// Set bash-completion.
+	_ = publishCmd.Flags().SetAnnotation("handler", cobra.BashCompSubdirsInDir, []string{})
+
+	faasCmd.AddCommand(publishCmd)
+}
+
+// publishCmd allows the user to build an OpenFaaS function container
+var publishCmd = &cobra.Command{
+	Use: `publish -f YAML_FILE [--no-cache] [--squash]
+  faas-cli publish --image IMAGE_NAME
+                   --handler HANDLER_DIR
+                   --name FUNCTION_NAME
+                   [--lang <ruby|python|python3|node|csharp|dockerfile>]
+                   [--no-cache] [--squash]
+                   [--regex "REGEX"]
+                   [--filter "WILDCARD"]
+                   [--parallel PARALLEL_DEPTH]
+                   [--build-arg KEY=VALUE]
+                   [--build-option VALUE]
+                   [--copy-extra PATH]
+                   [--tag <sha|branch|describe>]
+                   [--platforms linux/arm/v7]`,
+	Short: "Builds and pushes multi-arch OpenFaaS functions",
+	Long: `Builds and pushes multi-arch OpenFaaS functions using Docker buildx.
+A multi-arch template is required. Images will not be available in the 
+local Docker library. This command only works when a YAML file is specified.
+
+See also: faas-cli build`,
```

Comment:
Can we add `--platform` support to `faas-cli buid`? Note that it is singuler. 

It would be nice if there was some symmetry. 

There are technical limitations that mean it can only accept _one_ target platform, but the image would be available locally and could the be pushed independently. I think having `build` and `publish` might be confusing for users. Which should they use? I think it would be nice to have a simple answer like "you should default to using build for development and testing, if you function template supports multiarch then consider using publish`

Reply:
The way I see it:

* `publish` is for creating multi-arch image manifests from CI (`docker buildx` only)
* `build` and `up` are for local dev/test, and for when you have no need for multi-arch (`docker build` or shrinkwrap only)

The reason I was not keen on adding publish flags to build and push is that the way buildx works is distinct enough of a branch to need significant code changes, at the minimum a bunch of if statements.

Reply:
I've added notes in the README and the blurb for this concern.